src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+  (
+  -- * Data Types
+    Arity(..)
+  , Anf(..)
+  , App(..)
+  , Alloc(..)
+  , Rhs(..)
+
+  -- * Lowering to ANF
+  , toAnf
+  ) where
+
+import Hopper.Utils.LocallyNameless
+import Hopper.Internal.Core.Literal
+import Hopper.Internal.Core.Term
+
+import Data.Foldable (foldl')
+import Data.Maybe (fromMaybe, isJust)
+import Data.Monoid (Sum(..), Last(..))
+import Data.Word
+import Control.Arrow ((&&&), (***))
+import Control.Monad (replicateM, join)
+import Control.Monad.Trans.Reader (ReaderT, runReaderT, ask, local, reader)
+import Control.Monad.Trans.State.Strict (State, evalState)
+import Control.Lens hiding (levels)
+
+import qualified Data.Vector as V
+import qualified Data.Map.Strict as Map
+
+--------------------------
+-- Data Types
+--------------------------
+
+-- TODO: possibly allow more "atomic" expression types (besides variables) once
+--       we have some unboxed values
+
+-- TODO: switch back away from this
+newtype Arity = Arity Word32 deriving (Eq,Ord,Read,Show)
+
+data Anf
+  = AnfReturn !(V.Vector Variable) -- indices into the current env stack
+  | AnfLet !Arity -- TODO: switch back to !(V.Vector BinderInfo)
+           -- !(Maybe SourcePos)
+           !Rhs
+           !Anf
+  | AnfTailCall !App
+  deriving (Eq,Ord,Read,Show)
+
+data App
+  = AppFun !Variable
+           !(V.Vector Variable)
+  | AppPrim !PrimOpId !(V.Vector Variable)
+  | AppThunk !Variable
+  deriving (Eq,Ord,Read,Show)
+
+data Alloc
+  = AllocLit !Literal
+  | AllocLam !Arity -- TODO: switch back to !(V.Vector BinderInfo)
+             !Anf
+  | AllocThunk !Anf
+  deriving (Eq,Ord,Read,Show)
+
+data Rhs
+  = RhsAlloc !Alloc
+  | RhsApp !App
+  deriving (Eq,Ord,Read,Show)
+
+--------------------------
+-- Lowering to ANF
+--------------------------
+
+-- | The first slot in the most recent binder
+v0 :: Variable
+v0 = LocalVar $ LocalNamelessVar 0 $ BinderSlot 0
+
+-- | A temporary function to produce an 'Arity' for our ANF binding forms until
+-- we switch to using 'BinderInfo's in 'Anf' as well.
+arity :: V.Vector BinderInfo -> Arity
+arity binderInfos = Arity $ fromIntegral $ V.length binderInfos
+
+-- | A linear (single-use) reference to a(n existing/term or new/anf) binder. If
+-- two source variables refer to the same binder, they are tracked via separate
+-- 'AnfRef's.
+newtype AnfRef
+  = AnfRef { _refId :: Word64 }
+  deriving (Eq, Show, Ord)
+
+makeLenses ''AnfRef
+
+instance Enum AnfRef where
+  toEnum = AnfRef . toEnum
+  fromEnum = fromEnum . _refId
+
+-- | Bookkeeping structure that corresponds to a binder scope in the 'Term' AST.
+data BindingLevel
+  = BindingLevel { _levelRefs :: Map.Map AnfRef Variable
+                 -- ^ Linear references (to both existing/term and new/anf
+                 -- binders) to the 'Variable's they will be realized as when
+                 -- used. These variables are bumped as new 'AnfLet's are
+                 -- introduced.
+                 , _levelIntros :: Word32
```

Comment:
This comment could be clarified: "_number of binders_ introduced since the last source binder"